### PR TITLE
feat: only read iterator as far as necessary

### LINKED ISSUE #123: Table consumes iterables multiple times
I noticed that `Table` consumes the full iterator before rendering the rows. I would expect that the iterator gets consumed only once and only while the user scrolls in the table. The problem with the current behavior is that iterators sometimes cannot rewind or are expensive to create again. 

If I understand https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols correctly, it should be okay to implement an iterator where `[Symbol.iterator]` just returns the same iterator for an object.

Screen Shot 2021-03-22 at 12 24 32

See the problem in https://observablehq.com/d/d91b4879db425ea9

### PR DESCRIPTION
**Pull request by domoritz**
fixes #123

### CODE REVIEW

**Comment by domoritz** [comment]
I see @j-f1 implemented https://github.com/observablehq/inputs/pull/58 with a similar approach. You can choose whether you prefer this or the other approach.

File: src/table.js [hunk 1]
```diff
@@ -45,17 +40,53 @@ export function Table(
...
+    if (next == null) {
+      next = iterator.next;
+    }
+
+    if (length <= iterindex) {
+      return length;
+    }
+
+    while (length > iterindex) {
+      const {done, value} = next();
```
Comment (mbostock) [inline src/table.js:63]:
```suggestion
      const {done, value} = iterator.next();
```

Can we do this instead of stashing `next`?

**Comment by mbostock** [comment]
Thanks for the fix! I’ll make some style tweaks but this looks good.

File: src/table.js (hunk 1 above)
Comment (domoritz) [reply src/table.js:63]: Done

**Comment by domoritz** [comment]
Btw, here is the example I used to try the table (the dev setup is great!).

```html
import {Table} from "@observablehq/inputs";

const end = 1000;
const seekable = true;

const iter = {
  [Symbol.iterator]() {
    let i = 0;
    return {
      next: () => {
        if (i >= end) {
          console.log('Done');
          return { done: true };
        }
        const value = { i: i };

        if (i % 250 === 0) {
          console.log(i);
        }

        i += 1;
        return { value };
      }
    };
  },
  columns: ['i']
}
document.body.appendChild(Table(iter));
```

**Comment by domoritz** [comment]
Note that before the refactoring pull request, I had a version where we don't have to materialize the data in `array` but I used two iterators. This approach felt a bit cleaner but means we create an array (but only if we don't know the overall length of the iterator so that seems like a reasonable tradeoff).

**Comment by domoritz** [comment]
@mbostock If you can make time, it would be great if you could merge this fix.

**Comment by mbostock** [comment]
It’s on my list, just been busy. I’ll get to it.

**Pull request merged**